Refactoring compas.robots

gonzalocasas opened the pull request:
The main change in this PR is the replacement of `compas.robots` geometry primitives (eg `Origin`, `Box`, etc) with the ones from `compas.geometry`. I've kept backwards-compat deprecated aliases for everything thou.

### What type of change is this?

- [x] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

gonzalocasas committed:
fix error message for unsupported joint types

gonzalocasas committed:
allow non-standard attributes on URDF limit

gonzalocasas committed:
allow non-standard attributes on URDF mesh geometry

gonzalocasas committed:
Fix ctor defaults on RobotModel and remove redundant from_data

gonzalocasas committed:
fix data serialization for  materials without color

gonzalocasas committed:
Remove usage of deprecated color utilities

gonzalocasas committed:
Move some basic robot stuff to a base module

gonzalocasas committed:
Replace `Origin` for auto proxied `Frame` instances

gonzalocasas committed:
Avoid mutable list defaults

gonzalocasas committed:
Add proxy support directly to the URDF parser

gonzalocasas committed:
make ironpython happy

gonzalocasas committed:
correct param name

gonzalocasas committed:
Replace geometry primitives of compas.robots with compas.geometry ones

During parsing, the primitives are wrapped into proxies to add URDF-related features

gonzalocasas committed:
Update CHANGELOG

gonzalocasas committed:
Merge branch 'main' into docs-robots

gonzalocasas committed:
lint

gonzalocasas committed:
doc

gonzalocasas committed:
ah

gonzalocasas committed:
Document deprecated classes

gonzalocasas committed:
Update robots tutorial

gonzalocasas committed:
docs

tomvanmele commented on src/compas/robots/model/base.py:
```diff
@@ -0,0 +1,116 @@
+        return URDFElement('origin', attributes)
+
+    @classmethod
+    def from_urdf(cls, attributes, elements, text):
```
why not `text=None`?

gonzalocasas committed:
Forward __str__ to proxied object

tomvanmele commented on src/compas/robots/model/base.py:
```diff
@@ -0,0 +1,116 @@
+            Attributes of the URDF element.
+        elements: list[object]
+            Children elements of the URDF element.
+        text: str | None
```
if `text=None` then this would be `text : str, optional`

gonzalocasas replied:
;) yep, of course 🤦

tomvanmele commented on src/compas/robots/model/geometry.py:
```diff
@@ -3,28 +3,31 @@
 
+
 # Copied from https://github.com/ubernostrum/webcolors/blob/master/webcolors.py
 HTML4_NAMES_TO_HEX = {
```
perhaps this is something to be included in the color class and maps (not now of course)

gonzalocasas replied:
Added it to Color (https://github.com/compas-dev/compas/pull/983/commits/bb86ac9cc1f469d7954817e73162e27c0455bbf8)

tomvanmele commented on src/compas/robots/model/geometry.py:
```diff
@@ -46,369 +49,124 @@
+    # def from_data(cls, data):
+    #     box = cls('1 1 1')
+    #     box.data = data
+    #     return box
```
remove?

tomvanmele commented on src/compas/robots/model/geometry.py:
```diff
@@ -46,369 +49,124 @@
+#     def from_data(cls, data):
+#         cyl = cls(data['radius'], data['length'])
+#         cyl.data = data
+#         return cyl
```
same?

tomvanmele commented on src/compas/robots/model/geometry.py:
```diff
@@ -424,19 +182,21 @@ class MeshDescriptor(BaseShape):
         The scale factors of the mesh in the x-, y-, and z-direction.
-    geometry : :class:`compas.datastructures.Mesh`
-        The COMPAS geometry of this mesh.
+    meshes : list of :class:`compas.datastructures.Mesh`
```
```
list[:class:`~compas.datastructures.Mesh`]
```

tomvanmele commented on src/compas/robots/model/geometry.py:
```diff
@@ -424,19 +182,21 @@ class MeshDescriptor(BaseShape):
     """
 
-    def __init__(self, filename, scale='1.0 1.0 1.0'):
+    def __init__(self, filename, scale='1.0 1.0 1.0', **kwargs):
```
the docstring seems to be out of sync with the actual params

gonzalocasas replied:
fixed https://github.com/compas-dev/compas/pull/983/commits/4e1cc183d0fe7424290f1e8c495e1ac47ac15e20

tomvanmele commented on src/compas/robots/model/geometry.py:
```diff
@@ -623,34 +383,52 @@ def get_color(self):
-                r, g, b = hex_to_rgb(HTML4_NAMES_TO_HEX[self.name])
-                return [r / 255., g / 255., b / 255., 1.]
+                color = compas.colors.Color.from_hex(HTML4_NAMES_TO_HEX[self.name])
+                return color.r, color.g, color.b, color.a
```
i think the color class has a property `rgba`

gonzalocasas replied:
now it does https://github.com/compas-dev/compas/pull/983/commits/3a4a7816ee30dbc3eb07e01796700dde502cfc60

tomvanmele commented on src/compas/robots/model/geometry.py:
```diff
@@ -659,16 +437,14 @@ class Geometry(Data):
     Attributes
     ----------
-    shape : :class:`BaseShape`
+    shape : obj
```
`object`

tomvanmele commented on src/compas/robots/model/link.py:
```diff
@@ -2,49 +2,56 @@
+        shape = BoxProxy(primitive)
+        origin = Frame(*primitive.frame)
+        geometry = Geometry(box=shape)
+    if isinstance(primitive, Capsule):
```
`elif`?

tomvanmele commented on src/compas/robots/model/link.py:
```diff
@@ -2,49 +2,56 @@
-    if not shape:
+        origin = Frame.from_plane(plane)
+        geometry = Geometry(capsule=shape)
+    if isinstance(primitive, Cylinder):
```
`elif`?

tomvanmele commented on src/compas/robots/model/link.py:
```diff
@@ -2,49 +2,56 @@
+        shape = CylinderProxy(primitive)
+        origin = Frame.from_plane(primitive.circle.plane)
+        geometry = Geometry(cylinder=shape)
+    if isinstance(primitive, Sphere):
```
`elif`?

tomvanmele commented on src/compas/robots/model/link.py:
```diff
@@ -2,49 +2,56 @@
+        shape = SphereProxy(primitive)
+        origin = Frame(primitive.point, [1, 0, 0], [0, 1, 0])
+        geometry = Geometry(sphere=shape)
+    if not geometry:
```
`else`?

gonzalocasas replied:
I removed the entire thing now

tomvanmele commented on src/compas/robots/model/robot.py:
```diff
@@ -59,12 +58,12 @@ class RobotModel(Data):
     """
 
-    def __init__(self, name, joints=[], links=[], materials=[], **kwargs):
+    def __init__(self, name=None, joints=(), links=(), materials=(), **kwargs):
```
i think these can all be `None`

gonzalocasas replied:
These are actually magical. I don't know where to document this, but we rely on the default data types to do some introspection and determine the expected types during urdf parsing, so, I need these to be a sequence-type, but I changed from `[]` to `()` to avoid the problem of empty mutable lists in default params.

tomvanmele commented on src/compas/robots/model/robot.py:
```diff
@@ -59,12 +58,12 @@ class RobotModel(Data):
-        self.links = list(links)
-        self.materials = list(materials)
+        self.name = name or 'Robot'
+        self.joints = list(joints or [])
+        self.links = list(links or [])
+        self.materials = list(materials or [])
```
behaviour is the same with `None` as default

tomvanmele reviewed (commented):
generally looks good. a few small things...

gonzalocasas committed:
fix docstring

gonzalocasas committed:
Fix usage of primitives in programatic robot creation

Using a primitive was extracting the origin of the primitive itself, and moving it to the element's (visual/collision) origin, for no clear reason. Now the code leaves it at the primitive level and the origin in the element defaults to worldXY

gonzalocasas committed:
Add rgba to `compas.colors.Color`

gonzalocasas committed:
Merge branch 'main' into docs-robots

gonzalocasas committed:
add docstring to kwargs usage

gonzalocasas committed:
docstring

gonzalocasas committed:
remove commented out methods

gonzalocasas committed:
fix docstring

gonzalocasas committed:
fix surface artist registration

gonzalocasas committed:
lint

gonzalocasas committed:
Add Color.from_html_name

gonzalocasas committed:
Case-insensitive html name support

gonzalocasas committed:
Deprecate load_mesh in favor of load_meshes

gonzalocasas committed:
Handle invalid color names a bit more gracefully

gonzalocasas committed:
lint

gonzalocasas commented:
Ok, now it is ready for review.. I fixed the findings plus a bunch of other issues here and there, and updated (some) docs and robots tutorial.

tomvanmele commented on src/compas/colors/color.py:
```diff
@@ -182,6 +184,18 @@ def rgb255(self):
+    @property
+    def rgba255(self):
+        r, g, b = self.rgb255
+        a = int(self.b * 255)
```
`self.a`, no?

gonzalocasas replied:
ups.. good catch, thx

tomvanmele commented on src/compas/colors/color.py:
```diff
@@ -456,6 +470,28 @@ def from_hex(cls, value):
         return cls(r / 255.0, g / 255.0, b / 255.0)
 
+    @classmethod
+    def from_html_name(cls, name):
```
perhaps this doesn't need to mention "html"?

gonzalocasas replied:
true, done

tomvanmele reviewed (approved):
LGTM!
one typo, and one suggestion...

gonzalocasas committed:
typo

gonzalocasas committed:
No need to specify it's html in the method name

gonzalocasas committed:
Update changelog

gonzalocasas committed:
ah right...
